pkg/trace/agent/agent.go

```diff
@@ -190,6 +190,132 @@ func (a *Agent) loop() {
 	}
 }
 
+// sanitizeSpan applies some additional sanitization, including obfuscation,
+// truncation, and adding metadata based on the configured tags.
+func (a *Agent) sanitizeSpan(span *pb.Span, chunk *pb.TraceChunk, clientComputedTopLevel bool) {
+	for k, v := range a.conf.GlobalTags {
+		if k == tagOrigin {
+			chunk.Origin = v
+		} else {
+			traceutil.SetMeta(span, k, v)
+		}
+	}
+	if a.ModifySpan != nil {
+		a.ModifySpan(span)
+	}
+	a.obfuscateSpan(span)
+	Truncate(span)
+	if clientComputedTopLevel {
+		traceutil.UpdateTracerTopLevel(span)
+	}
+}
+
+// setRootSpanTags sets up any necessary tags on the root span.
+func (a *Agent) setRootSpanTags(root *pb.Span) {
+	clientSampleRate := sampler.GetGlobalRate(root)
+	sampler.SetClientRate(root, clientSampleRate)
+
+	if ratelimiter := a.Receiver.RateLimiter; ratelimiter.Active() {
+		rate := ratelimiter.RealRate()
+		sampler.SetPreSampleRate(root, rate)
+	}
+}
+
+// transformChunk handles a single trace chunk, applying the necessary transformations and sanitization,
+// before making a sampling decision. It returns the transformed chunk.
+// If this function returns nil, then the chunk should not be sampled, and should be removed from
+// the payload being processed.
+func (a *Agent) transformChunk(chunk *pb.TraceChunk, p *api.Payload, now time.Time, ts *info.TagStats, statsInput *stats.Input, ss *writer.SampledChunks) *pb.TraceChunk {
+	if len(chunk.Spans) == 0 {
```

Comment:
I'm not happy with this refactoring.  I agree the original function has grown too large and could really use refactoring, but this seems to just move most of the body into another function. Then this body manipulates its many parameters during the loop iterations.

Instead of making the logic clearer, I think this only further obfuscates what's going on by separating the state near the loop from the logic here manipulating that state.

If we want to break this down into sensible chunks, I'm all for it.